src/Orleans.Transactions/DistributedTM/Implementation/TransactionParticipant.cs

```diff
@@ -0,0 +1,904 @@
+                                {
+                                    logger.Trace($"committed {entry.Timestamp:o}");
+                                }
+                                entry.PromiseForTA.TrySetResult(TransactionalStatus.Ok);
+                            });
+
+                            if (entry.WriteParticipants.Count > 1)
+                            {
+                                // after committing, we need to run a task to confirm and collect
+                                storageBatch.FollowUpAction(() =>
+                                {
+                                    confirmationTasks.Add(entry.TransactionId, entry);
+                                    confirmationWorker.Notify();
+                                });
+                            }
+                            else
+                            {
+                                // there are no remote write participants to notify, so we can finish it all in one shot
+                                storageBatch.Collect(entry.TransactionId);
+                            }
+                            break;
+                        }
+
+                    case CommitRole.RemoteCommit:
+                        {
+                            if (entry.ConfirmationResponsePromise == null)
+                            {
+                                // this is a read-only participant that has sent
+                                // its prepared message.
+                                // So we are really done and need not store or do anything.
+                            }
+                            else
+                            {
+                                // we must confirm in storage, and then respond to TM so it can collect
+                                storageBatch.Confirm(entry.SequenceNumber);
+                                storageBatch.FollowUpAction(() =>
+                                {
+                                    entry.ConfirmationResponsePromise.TrySetResult(true);
+                                });
+                            }
+
+                            break;
+                        }
+
+                    case CommitRole.ReadOnly:
+                        {
+                            // we are a participant of a read-only transaction. Must store timestamp and then respond.
+                            storageBatch.Read(entry.Timestamp);
+                            storageBatch.FollowUpAction(() =>
+                            {
+                                entry.PromiseForTA.TrySetResult(TransactionalStatus.Ok);
+                            });
+
+                            break;
+                        }
+
+                    default:
+                        {
+                            logger.Error(777, $"internal error: impossible case");
+                            throw new NotImplementedException();
```

Comment:
Same as other case statement, i'll stop noting these as comment applies to all case statements.